### PR TITLE
Rule S2068: detect hard-coded passwords in web.config files

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5427

### CODE REVIEW

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs [hunk 1]
```diff
@@ -43,10 +45,13 @@ public abstract class DoNotHardcodeCredentialsBase<TSyntaxKind> : ParameterLoadi
 
         private readonly IAnalyzerConfiguration configuration;
         private readonly DiagnosticDescriptor rule;
+        private readonly Regex validCredentialPattern = new(@"^\?|:\w+|\{\d+[^}]*\}|""|'$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:48]: I doubt it's a RFC. It's more likely from AppSec research and FPs, see #2870

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:48]: The `\{\d+[^}]*\}` part is supposed to match format string likes `{0}`, `{0:#,0}`, `{0:ddMMyyy}`. So it makes sense exactly as written (give me anything until the next `}`

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:48]: Two double quotes as C# verbatim string, regex is looking for single doublequote and single singlequote

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:48]: Password starting with `:` are from this test case, I have no idea what kind of framework it is. It was likely source of FPs from AppSec research
```
string query3 = "password=:param";
```

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs [hunk 2]
```diff
@@ -106,89 +119,116 @@ protected bool IsEnabled(AnalyzerOptions options)
...
+            }
+
+            var match = passwordValuePattern.Match(variableValue);
+            if (match.Success && !IsValidCredential(match.Groups["suffix"].Value))
+            {
+                credentialWordsFound.Add(match.Groups["credential"].Value);
+            }
+
+            // Rule was initially implemented with everything lower (which is wrong) so we have to force lower
+            // before reporting to avoid new issues to appear on SQ/SC.
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:188]: The comment is from 2018, maybe SQ behavior was different back then. I'm not brave enough to risk changing this before LTS

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs [hunk 3]
```diff
@@ -71,14 +76,21 @@ public string CredentialWords
 
         protected DoNotHardcodeCredentialsBase(IAnalyzerConfiguration configuration)
         {
+            const string Rfc3986_Unreserved = "-._~";  // Numbers and letters are embedded in regex itself without escaping
+            const string Rfc3986_Pct = "%";
+            const string Rfc3986_SubDelims = "!$&'()*+,;=";
+            const string UriPasswordSpecialCharacters = Rfc3986_Unreserved + Rfc3986_Pct + Rfc3986_SubDelims;
+            // See https://tools.ietf.org/html/rfc3986 Userinfo can contain groups: unreserved | pct-encoded | sub-delims
+            var uriUserInfoPart = @"[\w\d" + Regex.Escape(UriPasswordSpecialCharacters) + "]+";
+            uriUserInfoPattern = new Regex(@"\w+:\/\/(?<Login>" + uriUserInfoPart + "):(?<Password>" + uriUserInfoPart + ")@", RegexOptions.Compiled);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:85]: I don't understand the comment. Can you provide more detailed explanation?

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:85]: I did not really changed anything of this logic, I just moved it to different place

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/WebConfig/DoNotHardcodeCredentials/Valid/Web.config [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      =
+      "Server=localhost; Database=Test; User=SA; Password=Secret123" />
+    <add >
+    </add>
+    <!--
+      Commented is not supported
+      <add connectionString="Server=localhost; Database=Test; User=SA; Password=Secret123"/>
+    -->
+    <add name="a" connectionString="Server=localhost; Database=Test; User=SA; Password=Secret123"/> <!-- Noncompliant -->
+    <add name="a" connectionString="Server=localhost; Database=Test; User=SA; Password="/>          <!-- Compliant, should not raise on empty passwords -->
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/WebConfig/DoNotHardcodeCredentials/Valid/Web.config:19]: We don't know if the string we're inspecting is connection string or what kind of system it is trying to connect to

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/WebConfig/DoNotHardcodeCredentials/Valid/Web.config:19]: There other UTs goes to the what the rule finds, I've added them to CS test cases, as it's not related to web.config files

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:84]: I think this is a valid userinfo that does not match the regex:
`https://user:passwordWith:isOkay@example.org/` This is how I read the grammar in the RFC, but I could be mistaken. At least .Net thinks so too [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfsQJwAUFABM4zAKYDujAqslHQCIAFggQAHAM4gA9NNQS2yEGIgSJHAPbIMAdSgIhIKBIDyAawgBPfPjYAPCGDEAbNgDotAc2kCAlG+4FZABJGAAzDQpfAG4gA==)

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs]: Can you please add comments for the alternations? I still have my doubts about some of them, but we don't need to fix them in this PR:

* Any format specifier doesn't make sense for passwords. How is `{0:ddMMyyy}` ever used for a password? So allowing `{0}` should be enough.
* I see how `Password=""` or `Password=''` makes sense, but `Password='` seems wrong. I don't think that `""` is related to verbatim strings. I think it is for an empty password.

If you think the same way, you should create an issue and/or ping AppSec for their opinion.

**Review by martin-strecker-sonarsource — COMMENTED** [review]
LGTM. The test for EF6 connection string should be updated to test the interesting case of an `&quote;` as a delimiter.

**Review by martin-strecker-sonarsource — APPROVED** [review]

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs]: I'll leave it out of scope to focus on delivering this PR about web.configs, not rewriting the whole rule

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:84]: We don't need to extract the full password. So changing it will make the regex more expensive, without affecting the results.

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:84]: The problem is that the `:` is not in the `<Password>` group and because of the trailing `@`, the regex doesn't match at all. The password group must be `"[\w\d" + Regex.Escape(UriPasswordSpecialCharacters) + ":]+"`.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:84]: I'll try in another PR

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:84]: See #6199
